.github/workflows/check.yaml

```diff
@@ -37,11 +37,61 @@ jobs:
           yamllint .yamllint snap/snapcraft.yaml
 
   build:
-    runs-on: ubuntu-22.04
+    needs:
+      - lint
+    runs-on: ${{ matrix.runs-on }}
+    strategy:
+      fail-fast: false
+      matrix:
+        runs-on: [[ubuntu-22.04]]
     steps:
       - uses: actions/checkout@v4
         with:
           fetch-depth: 0  # Complete git history is required to generate the version from git tags.
 
       - name: Verify snap builds successfully
+        id: build
         uses: snapcore/action-build@v1
+
+      - name: Determine system architecture
+        run: echo "SYSTEM_ARCH=$(uname -m)" >> $GITHUB_ENV
+
+      - name: Upload the built snap
+        uses: actions/upload-artifact@v4
+        with:
+          name: snap_${{ env.SYSTEM_ARCH }}
+          path: ${{ steps.build.outputs.snap }}
+
+  func:
+    needs:
+      - build
+    runs-on: ${{ matrix.runs-on }}
+    strategy:
+      fail-fast: false
+      matrix:
+        runs-on: [[ubuntu-22.04]]
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0  # Complete git history is required to generate the version from git tags.
+
+      - name: Determine system architecture
+        run: echo "SYSTEM_ARCH=$(uname -m)" >> $GITHUB_ENV
+
+      - name: Download the built snap
+        uses: actions/download-artifact@v4
+        with:
+          name: snap_${{ env.SYSTEM_ARCH }}
+
+      - name: Set up Python 3.10
+        uses: actions/setup-python@v5
+        with:
+          python-version: "3.10"
+
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          python -m pip install 'tox<5'
+
+      - name: Run functional tests
+        run: tox -e func
```

Comment:
this repo doesn't have a tox file, or any functional tests, so maybe we can't use this template :/

Reply:
We can add func test, and just do an endpoint testing because we don't necessarily have smart devices in the CI environment.

Reply:
~~It's a prefect case that using tox as an interface will encounter~~

It's ok to create an exception for this case right now from my point of view.